runners/direct-java/src/main/java/org/apache/beam/runners/direct/CommittedResult.java

```diff
@@ -18,10 +18,10 @@
 package org.apache.beam.runners.direct;
 
 import com.google.auto.value.AutoValue;
+import java.util.Optional;
```

Comment:
One reason not to use this is that it is not `Serializable` and the Java committee has decided it never should be. I obviously disagree, but until then I don't think `java.util.Optional` should be used.

Reply:
There was a very healthy discussion about usage of `Optional` on flink mailing list: https://lists.apache.org/thread.html/f5f8ce92f94c9be6774340fbd7ae5e4afe07386b6765ad3cfb13aec0@%3Cdev.flink.apache.org%3E
General consensus is that `Optional` should not be used for fields, but only for return values of performance non-critical code. I very much agree with that, if `Optional` is not used as field, its serializability doesn't matter. Moreover, DirectRunner does not rely that heavily on serialization (more than validating, that the user code is serializable). Using standard instead of guava seems to be beneficial to me, without any negative impact.

Reply:
Yes, that is exactly the logic I am familiar with. We've also had healthy discussions in Beam and seemingly every project. I can't escape null! I share the "null was a mistake, never really made sense in the first place, and should be eliminated from languages" perspective. I believe basic good programming practice to program as though it does not actually exist, except when necessary for compatibility or performance. Using `@Nullable` and `@NonNull` with an extended type checker is a compromise when necessary.

I have not seen any actual evidence for the claim that it causes performance problems. I would easily believe it - It is true that Java is exceptionally inefficient w.r.t. trivial wrappers unless they are well-optimized. But on the thread you linked the only person who brought up real performance said that they could not detect any difference except when _serializing_ it in performance-sensitive paths. So perhaps Java isn't so inefficient at optimizing it away after all.

In any event, please make it a separate change. You can do a bulk change in the codebase if you like. At this point, I am just so tired of this issue which humans are still stuck on that I will let most anything through :-/

Reply:
Yes, I also think that the performance impact of creating one short lived instance should be negligible. I actually tried to measure it, but I didn't get any statistically significant difference in this case. Although, I have seen significant impact in other cases (getting rid of creating useless instance in one place of spark runner gave me about 20% performance gain, but that could have been very specific case). But that is probably not that significant to this PR. :-)

To conclude this, I will drop the commit where I changed the Optional in DirectRunner and start a discuss thread related to making the usage consistent across the code base. Currently there is approx 3:2 usage of java.util vs. guava.